### PR TITLE
Removed hacky switch/case, migrated to if/else and early return statements

### PR DESCRIPTION
**Pull request by peter-gribanov**

### CODE REVIEW

**Comment by peter-gribanov** [comment]
@guilhermeblanco @Ocramius I have no idea why it is done
https://github.com/doctrine/doctrine2/commit/5e3e8b39571491ece27d755d625fce100ab6ca39#commitcomment-17772924

File: lib/Doctrine/ORM/EntityManager.php [hunk 1]
```diff
@@ -835,21 +836,16 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
-
-            case ($conn instanceof Connection):
-                if ($eventManager !== null && $conn->getEventManager() !== $eventManager) {
-                     throw ORMException::mismatchedEventManager();
-                }
-                break;
-
-            default:
-                throw new \InvalidArgumentException("Invalid argument: " . $conn);
+        if (is_array($conn)) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: Can you segregate this entire block into its own private static method?

**Comment by peter-gribanov** [comment]
@Ocramius it's better?

File: lib/Doctrine/ORM/EntityManager.php [hunk 2]
```diff
@@ -835,24 +836,38 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+     *
+     * @return Connection
+     *
+     * @throws ORMException
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected static function createConnection($conn, Configuration $config, EventManager $eventManager = null)
+    {
+        if (is_array($conn)) {
+            $conn = DriverManager::getConnection(
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: return early

File: lib/Doctrine/ORM/EntityManager.php [hunk 3]
```diff
@@ -835,24 +836,38 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+                $conn, $config, ($eventManager ?: new EventManager())
+            );
+        } elseif ($conn instanceof Connection) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: Make this a `if (! $conn instanceof Connection) { throw`

File: lib/Doctrine/ORM/EntityManager.php [hunk 4]
```diff
@@ -835,24 +836,38 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
-                break;
+        return new EntityManager($conn, $config, $conn->getEventManager());
+    }
 
-            default:
-                throw new \InvalidArgumentException("Invalid argument: " . $conn);
+    /**
+     * Factory method to create Connection instances.
+     *
+     * @param mixed         $conn         An array with the connection parameters or an existing Connection instance.
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: `array|Connection`

File: lib/Doctrine/ORM/EntityManager.php [hunk 5]
```diff
@@ -835,24 +836,38 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected static function createConnection($conn, Configuration $config, EventManager $eventManager = null)
+    {
+        if (is_array($conn)) {
+            $conn = DriverManager::getConnection(
+                $conn, $config, ($eventManager ?: new EventManager())
+            );
+        } elseif ($conn instanceof Connection) {
+            if ($eventManager !== null && $conn->getEventManager() !== $eventManager) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: `no need for `!== null`

**Comment by peter-gribanov** [comment]
@Ocramius now?

File: lib/Doctrine/ORM/EntityManager.php [hunk 6]
```diff
@@ -835,24 +836,33 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+     *
+     * @throws \InvalidArgumentException
+     */
+    protected static function createConnection($conn, Configuration $config, EventManager $eventManager = null)
+    {
+        if (is_array($conn)) {
+            return DriverManager::getConnection(
+                $conn, $config, ($eventManager ?: new EventManager())
+            );
+        } elseif ( ! $conn instanceof Connection) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: `else` not needed, since you return early

File: lib/Doctrine/ORM/EntityManager.php [hunk 7]
```diff
@@ -835,24 +836,33 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+     *
+     * @param array|Connection $conn         An array with the connection parameters or an existing Connection instance.
+     * @param Configuration    $config       The Configuration instance to use.
+     * @param EventManager     $eventManager The EventManager instance to use.
+     *
+     * @return Connection
+     *
+     * @throws \InvalidArgumentException
+     */
+    protected static function createConnection($conn, Configuration $config, EventManager $eventManager = null)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: `$conn` does not need to be abbreviated, now that I noticed it

File: lib/Doctrine/ORM/EntityManager.php [hunk 8]
```diff
@@ -835,24 +836,33 @@ public static function create($conn, Configuration $config, EventManager $eventM
...
+    {
+        if (is_array($conn)) {
+            return DriverManager::getConnection(
+                $conn, $config, ($eventManager ?: new EventManager())
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: This syntax is not PSR-2-compliant: split over multiple rows or pack it into a single row

Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: No parentheses needed around `($eventManager ?: new EventManager())`

File: lib/Doctrine/ORM/EntityManager.php [hunk 9]
```diff
@@ -820,39 +821,48 @@ public function initializeObject($obj)
...
+        if (is_array($connection)) {
+            return DriverManager::getConnection($connection, $config, $eventManager ?: new EventManager());
+        }
+
+        if ( ! $connection instanceof Connection) {
+            throw new \InvalidArgumentException("Invalid argument: " . $connection);
         }
 
-        return new EntityManager($conn, $config, $conn->getEventManager());
+        return $connection;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: Checking whether `$conn->getEventManager() === $eventManager` is still required (sorry, previous comment may have been misleading)

**Comment by Ocramius** [comment]
@peter-gribanov thanks!

**Pull request merged**